File: tonner/examples/falling-particles.py
import datetime
import tonner
import matplotlib.pyplot as plt

engine = tonner.Engine()

a = engine.add_particle(mass=1.0)
b = engine.add_particle(mass=1.0, velocity=[0, 0, 10])

dt = datetime.timedelta(milliseconds=10)

for body in [a, b]:
    engine.add_force(body, [0, 0, -10])

pos_a = []
pos_b = []
for _ in range(200):
    engine.simulate(dt)
    pos_a.append(engine.position(a)[2])
    pos_b.append(engine.position(b)[2])

plt.plot(pos_a, label="a")
plt.plot(pos_b, label="b")
plt.grid()
plt.legend()
plt.show()

File: tonner/examples/falling-particles.rs
use std::time::Duration;

use glam::DVec3;
use tonner::{Engine, ParticleBuilder};

fn main() {
    env_logger::init();

    let mut engine = Engine::new();

    let a = ParticleBuilder::default().mass(1.0).build(&mut engine);
    let b = ParticleBuilder::default()
        .mass(1.0)
        .velocity(10.0 * DVec3::Z)
        .build(&mut engine);

    let dt = Duration::from_secs(1);

    for body in [a, b] {
        *engine.force_mut(body).unwrap() += 10.0 * DVec3::NEG_Z;
    }

    for _ in 0..10 {
        engine.simulate(dt);
        println!("a: {}", engine.position(a).unwrap());
        println!("b: {}", engine.position(b).unwrap());
    }
}
